src/Hazelcast.Net/Core/SocketExtensions.cs

```diff
@@ -27,22 +27,98 @@ namespace Hazelcast.Core
+        }
+
+        private static async Task ConnectAsync5(this Socket socket, EndPoint endPoint, int timeoutMilliseconds, CancellationToken cancellationToken)
         {
-            // this is the code that the runtime uses, unchanged
+            static void ConnectCallback(object sender, SocketAsyncEventArgs a)
+            {
+                var completionSource = (TaskCompletionSource<Socket>) a.UserToken;
+                if (a.SocketError == SocketError.Success)
+                    completionSource.TrySetResult(a.ConnectSocket);
+                else
+                    completionSource.TrySetException(new SocketException((int) a.SocketError));
+            }
+
+            var connected = new TaskCompletionSource<Socket>();
+            var e = new SocketAsyncEventArgs { UserToken = connected };
+            e.RemoteEndPoint = endPoint;
+            e.Completed += ConnectCallback;
+            var pending = socket.ConnectAsync(e);
+
+            Task task;
+            if (pending)
+            {
+                var cancel = new CancellationTokenSource();
+                var cancelled = Task.Delay(timeoutMilliseconds, cancel.Token);
+                var reg = cancellationToken.Register(() => cancel.Cancel());
+
+                task = await Task.WhenAny(connected.Task, cancelled).CfAwait();
+                await reg.DisposeAsync().CfAwait();
+                cancel.Cancel();
+                cancel.Dispose();
+                await cancelled.CfAwaitNoThrow();
+            }
+            else
+            {
+                task = connected.Task;
+                ConnectCallback(null, e);
+            }
+
+            e.Dispose();
+
+            if (task == connected.Task)
+            {
+                await task.CfAwait(); // throw if there is anything to throw (failed to connect...)
+                return;
+            }
+
+            // we *need* to await connected.Task even when it was not the task returned by WhenAny else
+            // it may leak unobserved exceptions, as ConnectCallback may be invoked even when the socket
+            // is being teared down - however, ConnectCallback may also *not* be invoked, so we try to
+            // set result to force the task to complete, making sure we don't hang waiting for it
+            connected.TrySetResult(null);
+            await connected.Task.CfAwaitNoThrow(); // don't leave exceptions unobserved
+
+            TryCloseAndDispose(socket);
+
+            if (cancellationToken.IsCancellationRequested)
+                throw new OperationCanceledException("The socket connection operation has been canceled.");
+
+            throw new TimeoutException("The socket connection operation has timed out.");
+        }
+
+        public static async Task ConnectAsync4(this Socket socket, EndPoint endPoint, int timeoutMilliseconds, CancellationToken cancellationToken)
+        {
+            // note that purely using this code in .NET 6 leaves an unobserved exception behind
```

Comment:
can you put an assert statement to guarantee this?

Reply:
I don't understand what you want to assert. Let's discuss it.